pkg/volume/util.go

```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	}
+	if len(pvc.Spec.Selector.MatchExpressions) < 1 {
+		return make([]sets.String, 0), fmt.Errorf("key(s), operator(s) and value(s) are missing in selector.matchExpressions")
+	}
+	capacity := 0
+	for _, item := range pvc.Spec.Selector.MatchExpressions {
+		if item.Key == key && item.Operator == operator && len(item.Values) > 0 {
+			capacity++
+		}
+	}
+	if capacity == 0 {
+		return make([]sets.String, 0), fmt.Errorf("operator %q for key %q not found in selector.matchExpressions", key, operator)
+	}
+
+	ret := make([]sets.String, 0, capacity)
+	index := 0
+	for _, item := range pvc.Spec.Selector.MatchExpressions {
+		if item.Key == key && item.Operator == operator && len(item.Values) > 0 {
+			ret = append(ret, make(sets.String))
+			for _, value := range item.Values {
+				ret[index].Insert(value)
+			}
+			index++
+		}
+	}
+	return ret, nil
+}
+
+// Regions2ZonesParams are parameters for the func Regions2Zones
+type Regions2ZonesParams struct {
+	// set of regions to be converted to a set of zones
+	Regions sets.String
+	// all currently available zones
+	AllAvailableZones sets.String
+	// a func that converts a zone to a region
+	Zone2region func(string) (string, error)
+}
+
+// Regions2Zones converts a set of regions to a set of zones
+func Regions2Zones(in Regions2ZonesParams) (sets.String, error) {
+	ret := make(sets.String)
+	for zone := range in.AllAvailableZones {
+		for region := range in.Regions {
+			if regionOfZone, err := in.Zone2region(zone); err != nil {
+				return make(sets.String), fmt.Errorf("failed to convert zone (%v) to a region: %v", zone, err.Error())
+			} else if strings.Compare(regionOfZone, region) == 0 {
+				ret.Insert(zone)
+				break
+			}
+		}
+	}
+	return ret, nil
+}
+
+// CalculateSetOfZonesParams are parameters for the func CalculateSetOfZones
+type CalculateSetOfZonesParams struct {
+	// PVC data structure that is used for the set of zones calculation
+	PVC *v1.PersistentVolumeClaim
+	// is parameter zone specified in the Storage Class?
+	IsSCZoneSpecified bool
+	// in case the parameter zone is specified StorageClassZone contains the value of this parameter, otherwise it is ""
+	StorageClassZones string
+	// a func that returns a set of all available zones
+	GetAllZones func() (sets.String, error)
+	// a func that converts a zone to a region
+	Zone2region func(string) (string, error)
+}
+
+// CalculateSetOfZones returns:
+// - either a set of zones resulting from currently available zones, allowed zones by an admin in the corresponding storage class and zones preferred by the user in the selector part of the PVC
+// - or an error in case the resulting set of zones is empty or another error occurred
+func CalculateSetOfZones(input CalculateSetOfZonesParams) (sets.String, error) {
+	var ret, allAvailableZones sets.String
+	var err error
+	emptySet := make(sets.String)
+	if allAvailableZones, err = input.GetAllZones(); err != nil {
```

Comment:
I think if the user has specified zones in their storage class, we should not look at GetAllZones.